src/tools/cephfs/DataScan.cc

```diff
@@ -2046,19 +2046,13 @@ int MetadataDriver::inject_with_backtrace(
       }
     }
 
-    if (!created_dirfrag) {
-      // If the parent dirfrag already existed, then stop traversing the
-      // backtrace: assume that the other ancestors already exist too.  This
-      // is an assumption rather than a truth, but it's a convenient way
-      // to avoid the risk of creating multiply-linked directories while
-      // injecting data.  If there are in fact missing ancestors, this
-      // should be fixed up using a separate tool scanning the metadata
-      // pool.
-      break;
-    } else {
-      // Proceed up the backtrace, creating parents
-      ino = parent_ino;
-    }
+    // N.B.: when the metadata pool has suffered a partial loss (like one PG), then
```

Comment:
Should this change be optional ? Considering this would check all the ancestors for existance, not sure how much it would add up to the recovery time ?

Reply:
That's a reasonable concern. We could optimize it by caching dirfrags that have been created/linked.

Reply:
I think addressing this will require caching results from `read_dentry` but I believe some cases we expect the result to change after modifying the dentry. So I don't feel confident changing this in this PR since it's a performance consideration.

Reply:
I recommend we create a tracker ticket for the performance optimization after this is merged.

Reply:
I agree. Its critical that we get the functionality fixed first.

Reply:
Agreed